pkg/trace: Refactoring

ajgajg1134 opened the pull request:
<!--
* New contributors are highly encouraged to read our
  [CONTRIBUTING](/CONTRIBUTING.md) documentation.
* The pull request:
  * Should only fix one issue or add one feature at a time.
  * Must update the test suite for the relevant functionality.
  * Should pass all status checks before being reviewed or merged.
* Commit titles should be prefixed with general area of pull request's change.

-->
### What does this PR do?
A few things:
* Adds a new test to the integration tests to better ensure that normalization and obfuscation occurs
* Renames a few fields and functions in the `Process` function to better align with what they are
* Refactor the sampling done in the `Process` function to better consolidate it
  * This helped reveal some unnecessary `Clone` operations we were doing which should result in slightly better real world performance. However initial benchmarking has been too noisy to identify a consistent performance improvement.
  * Analytics events and single span sampling have been altered so instead of modifying the trace chunk directly they instead return which spans they would like to sample, and the higher level `sample` will decide to use those spans or not.
* Remove `TestNoTagNoTouch` since it doesn't _actually_ copy or verify that the tags aren't modified by the function. And the new function signature makes it clear the underlying `GetSingleSpanSampledSpans` doesn't modify the incoming parameters. In place of this I improved the `TestTagCausesInPlaceFilterAndKeep` test to REALLY copy the original payload and ensure that internal spans are not modified.
 
<!--
* A brief description of the change being made with this pull request.
* If the description here cannot be expressed in a succinct form, consider
  opening multiple pull requests instead of a single one.
-->

### Motivation
Tech debt.

In more detail: The Process function is long, difficult to reason about, and just generally "a bit of a mess" having had many features grafted on over the years. Specifically modifying sampling code has been extremely difficult as trace chunks were mutated internally and sampling logic was spread both inside and outside the Process function with mutation occurring in unexpected places. 
<!--
* What inspired you to submit this pull request?
* Link any related GitHub issues or PRs here.
-->

### Additional Notes

<!--
* Anything else we should know when reviewing?
* Include benchmarking information here whenever possible.
* Include info about alternatives that were considered and why the proposed
  version was chosen.
-->

### Possible Drawbacks / Trade-offs
I _believe_ that this makes the code clearer and may show some minor performance benefits from the removal of all the `clone` calls, however by allowing the structure to mutate with calls it's possible that future changes make this difficult to maintain long-term.
<!--
* What are the possible side-effects or negative impacts of the code change?
-->

### Describe how to test/QA your changes
This _should_ be a true refactor, however as mentioned in the motivation a lot of this code is confusing and bug-prone. I've taken great care to verify unit tests and integration tests as I go ensuring that changes that I make are well-covered before making them. However given the importance of this area of the trace-agent I think it's valuable to do a few short QA tests: (Note I have done these tests locally against the PR here but given the changes here I think it's worth doing them again as part of the QA process, I'm happy to help with any setup questions)

- [ ] Smoke test with default sampling rules and rates, Send much greater than 10TPS and ensure that sampling occurs and we see APPROXIMATELY 10TPS
- [ ] Test with analytics events (turns out these are used a bunch even though they're deprecated 😭). Enable analytics events with `DD_APM_ANALYZED_SPANS=service_name|operation_name=0.5` replacing service_name and operation_name with the span details you want.
- [ ] Test with sampled spans, use span sampling rules in the tracer to mark individual spans in a trace to be kept (and mark the whole traces as user manual drop). Ensure that only the span sampled spans appear in the UI.

<!--
* Write here in detail or link to detailed instructions on how this change can
  be tested/QAd/validated, including any environment setup.
-->

### Reviewer's Checklist
<!--
* Authors can use this list as a reference to ensure that there are no problems
  during the review but the signing off is to be done by the reviewer(s).

Note: Adding GitHub labels is only possible for contributors with write access.
-->

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

ajgajg1134 committed:
Add an additional integration level test

ajgajg1134 committed:
pkg/trace: start refactoring, renames, moving things adding comments

ajgajg1134 committed:
Refactor sampling

ajgajg1134 committed:
More sampling fixes, fix analytics events tests

ajgajg1134 committed:
Dont run analytics events for manual dropped traces

ajgajg1134 committed:
pkg/trace: More comments explaining DroppedTrace

ajgajg1134 committed:
pkg/trace: better names, clearer logs

ajgajg1134 committed:
cmd/trace-agent: add span sampling integration test

ajgajg1134 committed:
cmd/trace-agent: copyright header

darccio reviewed (approved):
LGTM. Thinking about how to avoid the noisy benchmarks, we could introduce in the future a "sampling strategy" parameter, so we can decouple the sampler and use a noop in benchmarks to measure the processing overhead in isolation. The same would apply for the sampler itself, as we could also test and benchmark isolated.

ahmed-mez commented on pkg/trace/agent/agent.go:
```diff
@@ -484,11 +475,37 @@ func isManualUserDrop(priority sampler.SamplingPriority, pt *traceutil.Processed
+
+	var events []*pb.Span
+	if !skipAnalyticsEvents {
+		// We skip analytics events when a trace is marked as manual drop (aka priority -1)
+		// Note that we DON'T skip single span sampling. We only do this for historical
+		// reasons and analytics events are deprecated so hopefully this can all go away someday.
```
I think ideally this comment should be in `traceSampling` where we actually decide the value of `skipAnalyticsEvents`. Doesn't hurt to duplicate the comment as well and keep it in both functions if you prefer!

ahmed-mez commented on pkg/trace/agent/agent.go:
```diff
@@ -484,11 +475,37 @@ func isManualUserDrop(priority sampler.SamplingPriority, pt *traceutil.Processed
+	keep, skipAnalyticsEvents := a.traceSampling(now, ts, pt)
+
+	var events []*pb.Span
+	if !skipAnalyticsEvents {
```
This is really tiny nit pick, I'd find it easier to read if it was `if analyticsEventsEnabled` instead of `if !skipAnalyticsEvents`

ajgajg1134 replied:
my only concern with a name like `analyticsEventsEnabled` is it looks like a control on if the feature itself is enabled (and not a dynamic change based on the sampling decision of the current trace) 🤔 maybe `checkAnalyticsEvents` or something else?

ahmed-mez commented on pkg/trace/agent/agent.go:
```diff
@@ -484,11 +475,37 @@ func isManualUserDrop(priority sampler.SamplingPriority, pt *traceutil.Processed
+			// This is OK because SSS is a replacement for analytics events so both should not be configured
+			// And when analytics events are fully gone we can get rid of all this
+			pt.TraceChunk.Spans = events
+		} else if len(events) > 0 {
+			log.Warnf("Detected both analytics events AND single span sampling in the same trace. Single span sampling wins because App Analytics is deprecated.")
+		}
```
Is this behavior covered in a unit test?

ajgajg1134 replied:
it is now!

ahmed-mez commented on pkg/trace/traceutil/azure.go:
```diff
@@ -37,7 +37,7 @@ const (
 )
 
-var appServicesTags map[string]string
+var appServicesTags map[string]string //TODO: make this actually cache?
```
I'd rather remove the variable and move the `// TODO: add caching` in `GetAppServicesTags`

ahmed-mez reviewed (approved):
Few minor comments but LGTM! It's definitely better now!

ajgajg1134 committed:
pkg/trace: PR Comments
Add unit test for single span sampling AND analytics events simultaneously
Some renames, cleanup some comments
